File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
public int first(int[]nums,int target){
    int n=nums.length;
    int low=0;
    int high=n-1;
    int res=-1;

    while(low<=high){
        int guess=(high+low)/2;

        if(nums[guess]<target){
            low=guess+1;
        }
        else if(nums[guess]>target){
            high=guess-1;
        }
        else{
            res=guess;
            high=guess-1;
        }
    }
    return res;
}

public int last(int[]nums,int target){
    int n=nums.length;
    int low=0;
    int high=n-1;
    int res=-1;

    while(low<=high){
        int guess=(high+low)/2;

        if(nums[guess]<target){
            low=guess+1;
        }
        else if(nums[guess]>target){
            high=guess-1;
        }
        else{
            res=guess;
            low=guess+1;
        }
    }
    return res;
}
    public int[] searchRange(int[] nums, int target) {
        int firstIdx=first(nums,target);
        int lastIdx=last(nums,target);

        return new int[]{firstIdx,lastIdx};
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.go
func first(nums []int, target int) int{
    n:=len(nums)
    low:=0
    high:=n-1
    res:=-1

    for low<=high{
        guess:=(low+high)/2

        if(nums[guess]<target){
            low=guess+1
        }else if(nums[guess]>target){
            high=guess-1
        }else{
            res=guess
            high=guess-1
        }
    }
    return res
}

func last(nums []int, target int) int{
    n:=len(nums)
    low:=0
    high:=n-1
    res:=-1

    for low<=high{
        guess:=(low+high)/2

        if(nums[guess]<target){
            low=guess+1
        }else if(nums[guess]>target){
            high=guess-1
        }else{
            res=guess
            low=guess+1
        }
    }
    return res
}

func searchRange(nums []int, target int) []int {
    firstIdx:=first(nums,target)
    lastIdx:=last(nums,target)

    return []int{firstIdx,lastIdx}
}